# js-testing-fundamentals

## basic

```js
const sum = (a, b) => a - b;

const result = sum(3, 7);
const expected = 10;

if (result !== expected) {
  throw new Error(`${result} is not equal to ${expected}`);
}
```

sum 함수를 선언할 때 실수가 있어서 a + b를 a - b로 작성했다.

이 경우 Error가 발생한다.

```js
const sum = (a, b) => a + b;
const subtract = (a, b) => a - b;

let result, expected;

result = sum(3, 7);
expected = 10;

if (result !== expected) {
  throw new Error(`${result} is not equal ${expected}`);
}

result = subtract(7, 3);
expected = 4;

if (result !== expected) {
  throw new Error(`${result} is not equal to ${expected}`);
}
```

테스트를 하나 더 추가했다.

테스트 과정을 추상화하여 다음과 같이 중복을 제거할 수 있다.

```js
const sum = (a, b) => a + b;
const subtract = (a, b) => a - b;

let result, expected;

result = sum(3, 7);
expected = 10;

expect(result).toBe(expected);

result = subtract(7, 3);
expected = 4;

expect(result).toBe(expected);

function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`);
      }
    }
  }
}
```

expect 함수는 일반 테스팅 라이브러리처럼 assert 함수를 추가하여 사용할 수 있다.

```js
function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`);
      }
    },
    toEqual(expected) {},
    toBeGreaterThen(expected) {},
  }
}
```

현재 코드는 sum을 테스트하는 과정에서 에러가 발생하면 subtract를 테스트 하기 전에 프로그램이 종료된다.

테스트에 에러가 발생하는 것과 무관하게 모든 테스트가 동작하도록 하려면 다음과 같이 작성한다.

```js
const sum = (a, b) => a - b;
const subtract = (a, b) => a - b;

let result, expected;

result = sum(3, 7);
expected = 10;

function sumTest() {
  expect(result).toBe(expected);
}

test('sum test', sumTest);

result = subtract(7, 3);
expected = 4;

function subtractTest() {
  expect(result).toBe(expected);
}

test('subtract test', subtractTest);

function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`);
      }
    },
    toEqual(expected) {},
    toBeGreaterThen(expected) {},
  }
}

function test(text, callback) {
  try {
    callback();
    console.log(`✅ ${text}`);
  } catch (error) {
    console.error(`🚨${text}`);
    console.error(error);
  }
}
```

코드를 좀 더 다듬으면 다음과 같다.

```js
const sum = (a, b) => a - b;
const subtract = (a, b) => a - b;

test('sum test', () => {
  const result = sum(3, 7);
  const expected = 10;
  expect(result).toBe(expected);
});

test('subtract test', () => {
  const result = subtract(7, 3);
  const expected = 4;
  expect(result).toBe(expected);
});

function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`);
      }
    },
    toEqual(expected) {},
    toBeGreaterThen(expected) {},
  }
}

function test(text, callback) {
  try {
    callback();
    console.log(`✅ ${text}`);
  } catch (error) {
    console.error(`🚨 ${text}`);
    console.error(error);
  }
}
```

async도 테스트 할 수 있어야 한다.

```js
const sum = (a, b) => a - b;
const subtract = (a, b) => a - b;

const sumAsync = (a, b) => Promise.resolve(sum(a, b));
const subtractAsync = (a, b) => Promise.resolve(subtract(a, b));

function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`);
      }
    }
  }
}

async function test(text, callback) {
  try {
    await callback();
    console.log(`✅ ${text}`);
  } catch (error) {
    console.error(`🚨 ${text}`);
    console.error(error);
  }
}

test('sum test', () => {
  const result = sum(3, 7);
  const expected = 10;
  expect(result).toBe(expected);
});

test('subtract test', () => {
  const result = subtract(7, 4);
  const expected = 3;
  expect(result).toBe(expected);
});

test('sum async test', async () => {
  const result = await sumAsync(3, 7);
  const expected = 10;
  expect(result).toBe(expected);
})
```

sumAsync가 비동기 함수이고 promise를 반환한다.

따라서 test에서 테스트 대상 함수(sumAsync)를 호출할 때 async, await를 추가한다.

그리고 test 함수 선언에서도 async와 await를 추가해야만 한다.

## global로 선언

test, expect 함수는 여러 테스트 파일에서 사용될 것이다.

그러므로 두 함수는 global로 선언하면 사용하기 편리하다.

root 경로에 setup-globals.js 라는 파일을 생성하고 test, expect 함수를 옮긴다.

```js
function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`);
      }
    }
  }
}

async function test(text, callback) {
  try {
    await callback();
    console.log(`✅ ${text}`);
  } catch (error) {
    console.error(`🚨${text}`);
    console.error(error);
  }
}

global.expect = expect;
global.test = test;
```

cli에서 node --require ./setup-globals.js simple-my.js 와 같이 실행하면 같은 파일에 해당 함수가 존재하는 것 처럼 실행된다.

이 과정은 jest를 사용하는 것과 유사하다.

jest를 설치하고 npx jest로 실행하면 simple.test.js 와 같이 파일명이 설정된 파일이 실행된다.
